File: fixtures/scale/java/medium/src/calc/lexer/Lexer.java
package calc.lexer;

import java.util.ArrayList;
import java.util.List;

import calc.errors.LexException;

public class Lexer {
    private final CharStream stream;

    public Lexer(String source) {
        this.stream = new CharStream(source);
    }

    public List<Token> tokenize() {
        List<Token> tokens = new ArrayList<>();
        stream.skipWhitespace();
        while (stream.hasNext()) {
            int at = stream.position();
            char c = stream.peek();
            if (Character.isDigit(c) || c == '.') {
                tokens.add(new Token(TokenType.NUMBER, stream.takeWhile(x -> Character.isDigit(x) || x == '.'), at));
            } else if (Character.isLetter(c)) {
                tokens.add(new Token(TokenType.IDENTIFIER, stream.takeWhile(Character::isLetterOrDigit), at));
            } else {
                TokenType type = Symbols.typeOf(stream.next());
                if (type == null) {
                    throw new LexException("unexpected character '" + c + "' at " + at);
                }
                tokens.add(new Token(type, String.valueOf(c), at));
            }
            stream.skipWhitespace();
        }
        tokens.add(new Token(TokenType.END, "", stream.position()));
        return tokens;
    }
}

File: fixtures/scale/dart/medium/lib/lexer/lexer.dart
import '../errors/lex_exception.dart';
import 'char_stream.dart';
import 'symbols.dart';
import 'token.dart';
import 'token_type.dart';

class Lexer {
  final CharStream _stream;

  Lexer(String source) : _stream = CharStream(source);

  List<Token> tokenize() {
    final tokens = <Token>[];
    _stream.skipWhitespace();
    while (_stream.hasNext) {
      final at = _stream.position;
      final c = _stream.peek();
      if (_isDigit(c) || c == '.') {
        tokens.add(Token(TokenType.number, _stream.takeWhile((x) => _isDigit(x) || x == '.'), at));
      } else if (_isLetter(c)) {
        tokens.add(Token(TokenType.identifier, _stream.takeWhile(_isLetterOrDigit), at));
      } else {
        final type = symbolTypeOf(_stream.next());
        if (type == null) throw LexException("unexpected character '$c' at $at");
        tokens.add(Token(type, c, at));
      }
      _stream.skipWhitespace();
    }
    tokens.add(Token(TokenType.end, '', _stream.position));
    return tokens;
  }

  bool _isDigit(String c) => c.compareTo('0') >= 0 && c.compareTo('9') <= 0;
  bool _isLetter(String c) => RegExp(r'[A-Za-z]').hasMatch(c);
  bool _isLetterOrDigit(String c) => _isDigit(c) || _isLetter(c);
}
